src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs

```diff
@@ -31,18 +30,21 @@ internal sealed class ServerMonitor : IServerMonitor
     {
         private static readonly TimeSpan __minHeartbeatInterval = TimeSpan.FromMilliseconds(500);
 
-        private readonly ExponentiallyWeightedMovingAverage _averageRoundTripTimeCalculator = new ExponentiallyWeightedMovingAverage(0.2);
         private readonly ServerDescription _baseDescription;
-        private readonly CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
+        private readonly CancellationTokenSource _cancellationTokenSource;
         private volatile IConnection _connection;
         private readonly IConnectionFactory _connectionFactory;
+        private volatile bool _currentCheckCancelled;
         private ServerDescription _currentDescription;
         private readonly EndPoint _endPoint;
+        private volatile BuildInfoResult _handshakeBuildInfoResult;
```

Comment:
I echo my previous concerns regarding using `volatile` here as well. Would a lock provide safer guarantees that are more easily reasoned about at a very very minor performance hit?